rules/firestore.json

```diff
@@ -45,40 +51,132 @@ service cloud.firestore {
+            // For any other fields, both players can modify them
+            allow update: if userIsVerified() &&
+                             preservesField('typeGame') &&
+                             (preservesField('playerZero') || resource.data.beginning == null) &&
+                             (preservesField('playerOne') || resource.data.beginning == null) &&
+                             (preservesField('beginning') || resource.data.beginning == null) &&
+                             (resource.data.playerZero == usernameOf(request.auth.uid) ||
+                              resource.data.playerOne == usernameOf(request.auth.uid));
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function notObservingOrTimedOut(creator, partId) {
+            let creatorUser =  get(/databases/$(database)/documents/joueurs/$(creator.id));
+            let notObserving = creatorUser.data.observedPart != partId;
+            let timedOut = duration.value(request.time.toMillis(), 'ms') > duration.value(creatorUser.data.last_changed.seconds, 's') + duration.value(10, 's');
+            return notObserving || timedOut;
+        }
+        match /joiners/{partId} {
+            // Anybody can create a joiner as long as they are themselves creator and if the corresponding part exists
+            allow create: if userIsVerified() &&
+                             userIsCreator(request.resource.data) &&
+                             exists(/databases/$(database)/documents/parties/$(partId));
+            // The creator can delete a joiner
+            allow delete: if userIsVerified() && userIsCreator(resource.data);
+            // Anybody else can delete a joiner only if the creator is offline
+            allow delete: if userIsVerified() && notObservingOrTimedOut(creatorOf(partId), partId);
+            // The creator can change all fields except changing partStatus to STARTED (3)
+            allow update: if userIsVerified() && userIsCreator(resource.data) &&
+                             request.resource.data.partStatus != 3;
+            // For anybody else:
+            //   - the chosen opponent can remove itself from the corresponding field
+            //   - the chosen opponent can change the config fro from PROPOSED to STARTED
+            allow update: if userIsVerified() && !userIsCreator(resource.data) &&
+                             preservesField('creator') &&
+                             preservesField('firstPlayer') &&
+                             preservesField('partType') &&
+                             preservesField('maximalMoveDuration') &&
+                             preservesField('totalPartDuration') &&
+                             (preservesField('partStatus') || changePartStatusAsChosenOpponent()) &&
+                             (preservesField('chosenOpponent') || removesOneSelfFromChosenPlayer());
```

Comment:
p-e mettre ces deux choix genre allow if (isChosenOpponent and legalChosenOpponentModification) or (removeOneSelf) ?